packages/editor/src/components/block-list/style.scss

```diff
@@ -820,39 +845,50 @@
 	}
 
 	.editor-block-list__block[data-align="left"] & {
-		@include break-small() {
-			// RTL note: this rule should not be auto-flipped based on direction.
-			/*rtl:ignore*/
-			float: left;
-		}
+		// RTL note: this rule should not be auto-flipped based on direction.
+		/*rtl:ignore*/
+		float: left;
 	}
 
 	.editor-block-list__block[data-align="right"] & {
-		@include break-small() {
-			// RTL note: this rule should not be auto-flipped based on direction.
-			/*rtl:ignore*/
-			float: right;
-		}
+		// RTL note: this rule should not be auto-flipped based on direction.
+		/*rtl:ignore*/
+		float: right;
+	}
+
+	.editor-block-list__block[data-align="left"] &,
+	.editor-block-list__block[data-align="right"] & {
+		// Move the block toolbar out of the flow using translate, but less for floats.
+		transform: translateY(-$block-padding -$border-width);
 	}
 }
 
 // Position the block toolbar when contextual.
 .editor-block-contextual-toolbar .editor-block-toolbar {
 	width: 100%;
 
-	// Hide right border on desktop, where the .components-toolbar instead has a right border.
 	@include break-small() {
+		width: auto;
+
+		// Hide right border on desktop, where the .components-toolbar instead has a right border.
 		border-right: none;
-	}
 
-	// This prevents floats from messing up the position.
-	@include break-small() {
+		// This prevents floats from messing up the position.
 		position: absolute;
 		left: 0;
 	}
+}
 
-	@include break-small() {
-		width: auto;
+// This rule ensures that any blocks that are not a Paragraph, that follows any aligned block, clears them.
+// This is necessary to ensure the selected block outlines and toolbar are correctly positioned.
+.editor-block-list__block[data-align] + .editor-block-list__block:not([data-align="right"]):not([data-type="core/paragraph"]) {
```

Comment:
Why is the paragraph block special here?

Reply:
This is the most concerning part for me but I don't see anything special about the paragraph block. Heading blocks are similar, lists are similar, quotes are similar. We could also image custom text blocks... 

I don't really know what these rules are doing, but we should definitely try to do without any block targetting here.